gufe/storage/stagingserialization.py

```diff
@@ -0,0 +1,95 @@
+from .stagingdirectory import StagingPath
+
+
+class StagingPathSerialization:
+    # TODO: where should this go? I think maybe on the storage manager
+
+    # Serializing staging paths
+    # -------------------------
+    #
+    # Some important user stories to consider:
+    #
+    # 1. I am loading my results object, and I will want to use the
+    #    associated files. This should be transparent, regardless of where
+    #    the permanent storage is located.
+    # 2. I am loading my results object, but I do not need the large stored
+    #    files. I do not want to download them when they aren't needed.
+    # 3. My permanent storage was a directory on my file system, but I have
+    #    moved that directory (with use cases of (a) I moved the absolute
+    #    path; (b) it is at a different relative path with respect to my
+    #    pwd.
+    # 4. I'm working with files from two different permanent storages. I
+    #    need to be able to load from both in the same Python process.
+    #
+    # Outputs from a protocol may contain a :class:`.StagingPath`. Note that
+    # a :class:`.StagingPath` is inherently not a complete description of
+    # how to obtain the associated data: in almost every situation, there is
+    # some additional context required. This can include the credentials to
+    # an external server, or a base path that the file can be found relative
+    # to (which may have changed if the user moves it.) Because of this, we
+    # need to inject that context in to the deserialization.
+    #
+    # This object injects the relevant context, provided by the
+    # :class:`.StorageManager`. It creates a JSONSerializerDeserializer
+    # based on the one being used by gufe in this process, using all the
+    # installed codecs plus an additional codec specific to this context.
+    #
+    # User stories 1 and 2 are handled by the nature of the
+    # :class:`.StagingPath` object. The external file downloads as part of
+    # the ``__fspath__`` method. This means that when using the ``open``
+    # builtin, you will automatically download the file to a local staging
+    # directory. However, the reference to the file can exist in the results
+    # object without downloading the file.
+    #
+    # User stories 3 and 4 are handled by this
+    # :class:`.StagingPathSerialization` class. Story 3 is handled by
+    # allowing the appropriate context (in the form of a
+    # :class:`.StorageManager`) to be injected into the deserialization
+    # process. Story 4 can be handled by using more than one
+    # :class:`.StagingPathSerialization` context (associated with different
+    # :class:`.StorageManager` objects.
+
+    def __init__(self, manager):
+        self.manager = manager
+        self.codec = JSONCodec(
+            cls=StagingPath,
+            to_dict=self.to_dict,
+            from_dict=self.from_dict,
+        )
+        self.refresh_handler()
+
+    def refresh_handler(self):
```

Comment:
what's the need for the refresh_handler to be a method rather than just inlined into `__init__`?

or what's the harm in just adding to `JSONCodec` on the import of this module?

Reply:
> what's the need for the refresh_handler to be a method rather than just inlined into __init__?

In principle, the codecs in `JSON_HANDLER` can change. For example, an external package can register additional handlers. Since So we at least want the ability to refresh. In fact, I was thinking about whether the `encoder`/`decoder` properties maybe should call `refresh_handler`.

> or what's the harm in just adding to `JSONCodec` on the import of this module?

I'm not sure what you mean by "adding to `JSONCodec`." Could you clarify?

Reply:
Sorry I meant JSON_HANDLER.  Other Codecs are defined and just appended onto the HANDLER as they are defined, so the HANDLER is "batteries included".  Is there a reason this Codec can't have the same treatment?

Reply:
1. We don't know the storage paths at import time, so this is literally impossible.
2. We may have more than one set of `shared`/`permanent` (especially `permanent`) defined in a given process. Imagine that I'm working with data from multiple campaigns. Imagine furthermore that I've use the `keep_shared` option so that my shared stuff is also still accessible. This precludes any kind of serialization that relies on global state. [This test](https://github.com/OpenFreeEnergy/gufe/blob/e39a79026c35b0b9a8643e4f8c1001bd81b74cc3/gufe/tests/storage/test_stagingserialization.py#L198) is specifically to prevent us from making this mistake.